content/crutchcorn/collections/art-of-accessibility/posts/art-of-a11y-labels/index.md

```diff
@@ -0,0 +1,869 @@
+  />
+  <p class="errormessage" :id="props.id + '-error'">{{ props.error }}</p>
+</template>
+
+<style scoped>
+.label {
+  margin-right: 1rem;
+}
+
+.errormessage {
+  color: red;
+}
+</style>
+```
+
+```vue
+<!-- App.vue -->
+<script setup>
+import TextInput from "./TextInput.vue";
+</script>
+
+<form>
+    <TextInput label="Email" id="email" error="Invalid email" />
+    <TextInput label="Password" type="password" id="password" />
+    <button type="submit">Login</button>
+</form>
+```
+
+<!-- ::end:tabs -->
+
+Now we can see our form with a warning about an invalid email. It looks something like this when an invalid email is entered:
+
+----
+
+<form class="__form_error_example_1" onsubmit="event.preventDefault()"><label for="email" class="label">Email</label><input id="email" aria-invalid="true" aria-errormessage="email-error"><p class="errormessage" id="email-error">Invalid email</p><label for="password" class="label">Password</label><input id="password" type="password" aria-invalid="false" aria-errormessage="password-error"><p class="errormessage" id="password-error"></p><button type="submit">Login</button><style>
+.__form_error_example_1 .label {
+  margin-right: 1rem;
+}
+.__form_error_example_1 .errormessage {
+  color: red;
+}</style></form>
+
+
+---
+
+## Generating Unique IDs Automatically using UUIDv4 
+
+Our forms above are pretty functional now, but there's a small developer experience headache associated with our new `TextInput` form: You are _required_ to define an unique `id` manually for each field.
+
+While this isn't a problem for small forms, as your application grows this can be quite a headache remembering all of the used `id`s for new forms.
+
+While it would be nice to remove the requirement to pass a custom `id`, we need to have one present in our input to link the `label`, error `span`, and `input` together. Luckily for us, computers have had a fairly reliable way of generating a unique IDs in the form of **UUIDs**.
+
+> What are UUIDs?
+
+While [I've written in depth about UUIDs before](/posts/what-are-uuids), the gist of "What is a UUID" is "They're a method of generating unique IDs for items using a specific algorithm, designated by a 'version' of UUID used to generate the ID".
+
+For our purposes, we'll be using a [UUIDv4](/posts/what-are-uuids#UUIDv4) to generate truly unique IDs for each DOM element we want to associate.
+
+Since there's not a method for generating UUIDs as part of JavaScript's core APIs, let's install a `uuid` package to do this for us:
+
+```
+npm install uuid
+```
+
+Now that we have the ability to generate UUIDv4s using:
+
+```javascript
+import {v4 as uuidv4} from 'uuid';
+
+uuidv4();
+```
+
+Let's look at how we can integrate this into our `TextInput` component:
+
+<!-- ::start:tabs -->
+
+### React
+
+```jsx {5}
+// TextInput.jsx
+import {v4 as uuidv4} from 'uuid';
+
+export const TextInput = ({ label, type, id, error }) => {
+	const [uuid] = useState(uuidv4());
```

Comment:
I'm curious why the use of `useState()` instead of `useRef()`. Admittedly, I don't know which is better here since I use the React built-in `useId()` which is dependency free and works with React Server Components (currently non-async only, apparently), unlike the client-only hooks.

Reply:
Oof, I forgot that both React and Vue have `useId`. I'll rewrite this